Mathlib/Algebra/Pointwise/Stabilizer.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.Finset.Pointwise
+import Mathlib.GroupTheory.QuotientGroup
+
+/-!
+# Stabilizer of a set under a pointwise action
+
+This file characterises the stabilizer of a set/finset under the pointwise action of a group.
+-/
+
+open Function Set
+open scoped Pointwise
+
+namespace MulAction
+variable {G α : Type*} [Group G] [MulAction G α] {a : G}
+
+section Set
+
+@[to_additive (attr := simp)]
+lemma stabilizer_empty : stabilizer G (∅ : Set α) = ⊤ :=
+  Subgroup.coe_eq_univ.1 $ eq_univ_of_forall fun _a ↦ smul_set_empty
+
+@[to_additive (attr := simp)]
+lemma stabilizer_singleton (b : α) : stabilizer G ({b} : Set α) = stabilizer G b := by ext; simp
+
+@[to_additive]
+lemma mem_stabilizer_set {s : Set α} : a ∈ stabilizer G s ↔ ∀ b, a • b ∈ s ↔ b ∈ s := by
+  refine ⟨fun h b ↦ ?_, fun h ↦ ?_⟩
+  · rw [← (smul_mem_smul_set_iff : a • b ∈ a • s ↔ _), mem_stabilizer_iff.1 h]
+  simp_rw [mem_stabilizer_iff, Set.ext_iff, mem_smul_set_iff_inv_smul_mem]
+  exact ((MulAction.toPerm a).forall_congr' $ by simp [Iff.comm]).1 h
+
+@[to_additive (attr := simp)]
+lemma stabilizer_subgroup (s : Subgroup G) : stabilizer G (s : Set G) = s := by
+  simp_rw [SetLike.ext_iff, mem_stabilizer_set]
+  refine fun a ↦ ⟨fun h ↦ ?_, fun ha b ↦ s.mul_mem_cancel_left ha⟩
+  simpa only [smul_eq_mul, SetLike.mem_coe, mul_one] using (h 1).2 s.one_mem
+
+@[to_additive]
+lemma map_stabilizer_le (f : G →* G) (s : Set G) :
+    (stabilizer G s).map f ≤ stabilizer G (f '' s) := by
+  rintro a
+  simp only [Subgroup.mem_map, mem_stabilizer_iff, exists_prop, forall_exists_index, and_imp]
+  rintro a ha rfl
+  rw [← image_smul_distrib, ha]
+
+@[to_additive (attr := simp)]
+lemma stabilizer_mul_self (s : Set G) : (stabilizer G s : Set G) * s = s := by
+  ext
+  refine ⟨?_, fun h ↦ ⟨_, _, (stabilizer G s).one_mem, h, one_mul _⟩⟩
+  rintro ⟨a, b, ha, hb, rfl⟩
+  rw [← mem_stabilizer_iff.1 ha]
+  exact smul_mem_smul_set hb
+
+end Set
```

Comment:
Can you add some subheadings for `Set` and `Finset`?

Reply:
Sure. Btw I really dislike how I have to `MulAction.something_stabilizer_set`/`MulAction.something_stabilizer_finset`. I would much prefer being able to write `Set.something_mulStabilizer`/`Finset.something_addStabilizer`. Do you think it's worth a rename?